sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -0,0 +1,303 @@
+          .add("instanceId", instanceId)
+          .add("databaseId", databaseId)
+          .toString();
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder
+          .addIfNotNull(DisplayData.item("projectId", projectId)
+              .withLabel("Output Project"))
+          .addIfNotNull(DisplayData.item("instanceId", instanceId)
+              .withLabel("Output Instance"))
+          .addIfNotNull(DisplayData.item("databaseId", databaseId)
+              .withLabel("Output Database"));
+    }
+
+    public String getProjectId() {
+      return projectId;
+    }
+
+    public String getInstanceId() {
+      return instanceId;
+    }
+
+    public String getDatabaseId() {
+      return databaseId;
+    }
+
+  }
+
+
+  /**
+   * {@link DoFn} that writes {@link Mutation}s to Cloud Spanner. Mutations are written in
+   * batches, where the maximum batch size is {@link SpannerIO#SPANNER_MUTATIONS_PER_COMMIT_LIMIT}.
+   *
+   * <p>See <a href="https://cloud.google.com/spanner"/>
+   *
+   * <p>Commits are non-transactional.  If a commit fails, it will be retried (up to
+   * {@link SpannerIO#SPANNER_MUTATIONS_PER_COMMIT_LIMIT}. times). This means that the
+   * mutation operation should be idempotent.
+   */
+  @VisibleForTesting
+  static class SpannerWriterFn<T, V> extends DoFn<Mutation, Void> {
+    private static final Logger LOG = LoggerFactory.getLogger(SpannerWriterFn.class);
+    private transient Spanner spanner;
+    private final String projectId;
+    private final String instanceId;
+    private final String databaseId;
+    private transient DatabaseClient dbClient;
+    // Current batch of mutations to be written.
+    private final List<Mutation> mutations = new ArrayList<Mutation>();
+
+    private static final int MAX_RETRIES = 5;
+    private static final FluentBackoff BUNDLE_WRITE_BACKOFF =
+        FluentBackoff.DEFAULT
+            .withMaxRetries(MAX_RETRIES).withInitialBackoff(Duration.standardSeconds(5));
+
+    @VisibleForTesting
+    SpannerWriterFn(String projectId, String instanceId, String databaseId) {
+      this.projectId = checkNotNull(projectId, "projectId");
+      this.instanceId = checkNotNull(instanceId, "instanceId");
+      this.databaseId = checkNotNull(databaseId, "databaseId");
+    }
+
+    @Setup
+    public void setup() throws Exception {
+        SpannerOptions options = SpannerOptions.newBuilder().build();
+        spanner = options.getService();
+    }
+
+    @StartBundle
+    public void startBundle(Context c) throws IOException {
+      dbClient = getDbClient(spanner, DatabaseId.of(projectId, instanceId, databaseId));
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) throws Exception {
+      Mutation m = c.element();
+      int columnCount = m.asMap().size();
+      if ((mutations.size() + 1) * columnCount >= SpannerIO.SPANNER_MUTATIONS_PER_COMMIT_LIMIT) {
```

Comment:
I wonder if the batch size needs to be configurable? In cases there are large mutations, you might want a smaller batch size.